File: 429.N-ary-Tree-Level-Order-Traversal.py
# https://leetcode.com/problems/n-ary-tree-level-order-traversal/description/
#
# algorithms
# Easy (54.41%)
# Total Accepted:    7.4k
# Total Submissions: 13.6k
# beats 16.17% of python submissions


"""
# Definition for a Node.
class Node(object):
    def __init__(self, val, children):
        self.val = val
        self.children = children
"""


class Solution(object):
    def levelOrder(self, root):
        """
        :type root: Node
        :rtype: List[List[int]]
        """
        if not root:
            return []

        res, level = [], []
        nodes = [root, '#']
        i = 0

        while i < len(nodes):
            node = nodes[i]
            if node == '#':
                res.append(level)
                level = []
                if i == len(nodes) - 1:
                    break
                else:
                    nodes += '#',
            else:
                level += node.val,
                if node.children:
                    nodes.extend(node.children)
            i += 1

        return res

File: 429.N-ary-Tree-Level-Order-Traversal.java
// https://leetcode.com/problems/n-ary-tree-level-order-traversal/
//
// algorithms
// Medium (61.87%)
// Total Accepted: 56,042
// Total Submissions: 90,585


/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
    public List<List<Integer>> levelOrder(Node root) {
        List<List<Integer>> res = new ArrayList<>();
        if (root == null) {
            return res;
        }

        LinkedList<Node> q = new LinkedList<>();
        q.offer(root);
        q.offer(null);
        List<Integer> level = new ArrayList<>();

        while (q.size() > 0) {
            Node cur = q.pollFirst();
            if (cur == null) {
                res.add(level);
                if (q.size() > 0) {
                    level = new ArrayList<>();
                    q.offer(null);
                }
            } else {
                level.add(cur.val);
                for (Node n : cur.children) {
                    q.offer(n);
                }
            }
        }

        return res;
    }
}